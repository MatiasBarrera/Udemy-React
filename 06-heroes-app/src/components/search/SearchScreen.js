import React, { useMemo } from 'react'
import queryString from 'query-string'
import { useHistory, useLocation } from 'react-router'
import useForm from '../../hooks/useForm'
import { HeroCard } from '../heroes/HeroCard'
import { getHeroesByName } from '../selectors/getHeroesByName'

export const SearchScreen = () => {
  const history = useHistory()
  const { search } = useLocation()

  const { q = '' } = queryString.parse(search)

  const [formValues, handleInputChange] = useForm({
    searchText: q
  })

  const { searchText } = formValues

  const heroesFiltered = useMemo(() => getHeroesByName(q), [q])
  // const heroesFiltered = getHeroesByName(searchText)

  const handleSearch = e => {
    e.preventDefault()
    history.push(`?q=${searchText}`)

    // reset()
  }

  return (
    <>
      <h1>Search Screen</h1>
      <div className="search-container">
        <div className="search-form">
          <h4>Search Form: </h4>
          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Find your hero"
              name="searchText"
              value={searchText}
              onChange={handleInputChange}
              autoComplete="off"
            />

            <button type="submit">🔍</button>
          </form>
        </div>
        <br />
        <h4>Results: </h4>
        <br />
        <div className="search-result">
          {q === '' && <div className="alert alert-info">Search a hero</div>}
          {q !== '' && heroesFiltered.length === 0 && (
            <div className="alert alert-danger">
              There is no a hero with: {q}
            </div>
          )}
          {heroesFiltered.map(hero => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  )
}
