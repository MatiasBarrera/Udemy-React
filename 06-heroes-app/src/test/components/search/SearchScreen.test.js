import { mount } from 'enzyme'
import React from 'react'
import { MemoryRouter, Route, Router } from 'react-router'
// import { AuthContext } from '../../../auth/AuthContext'
import { SearchScreen } from '../../../components/search/SearchScreen'

describe('Pruebas en SearchScreen', () => {
  test('Debe de mostrarse correctamente con valores por defecto', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero')
  })

  test('Debe de mostrar a batman y el input con el valor del queryString', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )

    expect(wrapper.find('input').prop('value')).toBe('batman')
    expect(wrapper.find('HeroCard').exists()).toBeTruthy()
  })

  test('Debe de mostrar error si el hero buscado no existe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    )
    expect(wrapper.find('input').prop('value')).toBe('batman123')
    expect(wrapper.find('.alert-danger').text().trim()).toBe(
      'There is no a hero with: batman123'
    )
    expect(wrapper).toMatchSnapshot()
  })

  test('Debe de llamar el push del history', () => {
    const historyMock = {
      push: jest.fn(),
      replace: jest.fn(),
      location: {
        hash: '',
        key: 've7stz',
        pathname: '/search',
        search: '?q=captain',
        state: undefined
      },
      listen: jest.fn(),
      createHref: jest.fn()
    }

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=captain']}>
        <Router history={historyMock}>
          <Route path="/search" component={() => <SearchScreen />} />
        </Router>
      </MemoryRouter>
    )

    wrapper.find('input').simulate('change', {
      target: {
        name: 'searchText',
        value: 'captain'
      }
    })
    wrapper.find('form').prop('onSubmit')({
      preventDefault() {}
    })

    expect(historyMock.push).toHaveBeenCalledWith('?q=captain')
    expect(wrapper).toMatchSnapshot()
  })
})
