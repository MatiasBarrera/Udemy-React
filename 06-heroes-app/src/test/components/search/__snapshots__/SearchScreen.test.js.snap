// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScreen Debe de llamar el push del history 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <div
    className="search-container"
  >
    <div
      className="search-form"
    >
      <h4>
        Search Form: 
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="captain"
        />
        <button
          type="submit"
        >
          üîç
        </button>
      </form>
    </div>
    <br />
    <h4>
      Results: 
    </h4>
    <br />
    <div
      className="search-result"
    >
      <div
        className="card animate__fadeIn animate__animated animate__fast"
      >
        <div>
          <img
            alt="Captain America"
            className="card-img"
            src=""
          />
        </div>
        <div
          className="card-body"
        >
          <h5>
            Captain America
          </h5>
          <p
            className="card-text"
          >
            Steve Rogers
          </p>
          <p
            className="card-text"
          >
            <small
              className="text-muted"
            >
              Captain America Comics #1
            </small>
          </p>
          <a
            className="link"
            onClick={[Function]}
          >
            Mas...
          </a>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en SearchScreen Debe de mostrar error si el hero buscado no existe 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <div
    className="search-container"
  >
    <div
      className="search-form"
    >
      <h4>
        Search Form: 
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman123"
        />
        <button
          type="submit"
        >
          üîç
        </button>
      </form>
    </div>
    <br />
    <h4>
      Results: 
    </h4>
    <br />
    <div
      className="search-result"
    >
      <div
        className="alert alert-danger"
      >
        There is no a hero with: 
        batman123
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en SearchScreen Debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <div
    className="search-container"
  >
    <div
      className="search-form"
    >
      <h4>
        Search Form: 
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          type="submit"
        >
          üîç
        </button>
      </form>
    </div>
    <br />
    <h4>
      Results: 
    </h4>
    <br />
    <div
      className="search-result"
    >
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>,
]
`;
